napari/_qt/dialogs/qt_package_installer.py

```diff
@@ -0,0 +1,454 @@
+"""
+A tool-agnostic installation logic for the plugin manager.
+
+The main object is `InstallerQueue`, a `QProcess` subclass
+with the notion of a job queue. The queued jobs are represented
+by a `deque` of `*InstallerTool` dataclasses that contain the
+executable path, arguments and environment modifications.
+Available actions for each tool are `install`, `uninstall`
+and `cancel`.
+"""
+import contextlib
+import os
+import shutil
+import sys
+from collections import deque
+from dataclasses import dataclass
+from enum import Enum
+from logging import getLogger
+from pathlib import Path
+from tempfile import gettempdir
+from typing import Deque, Optional, Sequence, Tuple
+
+from npe2 import PluginManager
+from qtpy.QtCore import QObject, QProcess, QProcessEnvironment, Signal
+from qtpy.QtWidgets import QTextEdit
+
+from napari._version import version as _napari_version
+from napari._version import version_tuple as _napari_version_tuple
+from napari.plugins import plugin_manager
+from napari.plugins.pypi import _user_agent
+from napari.utils._appdirs import user_plugin_dir, user_site_packages
+from napari.utils.misc import running_as_bundled_app
+from napari.utils.translations import trans
+
+JobId = int
+log = getLogger(__name__)
+
+
+class InstallerActions(str, Enum):
+    "Available actions for the plugin manager"
+    install = "install"
+    uninstall = "uninstall"
+    cancel = "cancel"
+
+
+class InstallerTools(str, Enum):
+    "Available tools for InstallerQueue jobs"
+    conda = "conda"
+    pip = "pip"
+
+
+@dataclass(frozen=True)
+class AbstractInstallerTool:
+    action: InstallerActions
+    pkgs: Tuple[str, ...]
+    origins: Tuple[str, ...] = ()
+    prefix: Optional[str] = None
+
+    @property
+    def ident(self):
+        return hash((self.action, *self.pkgs, *self.origins, self.prefix))
+
+    # abstract method
+    def executable(self):
+        raise NotImplementedError()
+
```

Comment:
why is it an instance method, not classmethod or staticmethod?

Reply:
Hm assuming it will never depend on any parameters, that could be, yes. Some tools might depend on `prefix`, but we haven't found that case yet. I'll make it a classmethod!